src/SingletonPaymasterV6.sol

```diff
@@ -192,24 +192,23 @@ contract SingletonPaymasterV6 is BaseSingletonPaymaster, IPaymasterV6 {
      * @param _actualGasCost The total gas cost (in wei) of this userOperation.
      */
     function _postOp(PostOpMode _postOpMode, bytes calldata _context, uint256 _actualGasCost) internal {
-        if (_postOpMode == PostOpMode.postOpReverted) {
-            return;
-        }
-
         ERC20PostOpContext memory ctx = _parsePostOpContext(_context);
 
         uint256 actualUserOpFeePerGas = _calculateActualUserOpFeePerGas(ctx.maxFeePerGas, ctx.maxPriorityFeePerGas);
 
-        uint256 costInToken = getCostInToken(
-            _actualGasCost / actualUserOpFeePerGas, ctx.postOpGas, actualUserOpFeePerGas, ctx.exchangeRate
-        ) + ctx.constantFee;
+        uint256 costInToken =
+            getCostInToken(_actualGasCost, ctx.postOpGas, actualUserOpFeePerGas, ctx.exchangeRate) + ctx.constantFee;
 
         // There is a bug in EntryPoint v0.6 where if postOp reverts where the revert bytes are less than 32bytes,
         // it will revert the whole bundle instead of just force failing the userOperation.
         // To avoid this we need to use `trySafeTransferFrom` to catch when it revert and throw a custom
         // revert with more than 32 bytes. More info: https://github.com/eth-infinitism/account-abstraction/pull/293
         bool success = SafeTransferLib.trySafeTransferFrom(ctx.token, ctx.sender, ctx.treasury, costInToken);
 
+        if (_postOpMode == PostOpMode.postOpReverted && !success) {
```

Comment:
Just a thought, I am wondering if we should emit a log for this so it will be easy to index and use in offchain logic